### PR TITLE
CAD-2167: TraceForwarder with a callback.

### PR DESCRIPTION
**Pull request by denisshevchenko**
The node has important information about itself:

1. Protocol
2. Version
3. Commit
4. Node start time
5. Slot length
6. Epoch length
7. KES period length
8. System start time

All this information should be traced to RTView (by `TraceForwarderBK` plugin) automatically, once after each connection with TraceAcceptorBK is established. Now `TraceForwarderBK` plugin provides a callback function for it.

_IMPORANT: Now it doesn't compile, I need help with Byron start time._

### CODE REVIEW

File: cardano-node/src/Cardano/Node/Configuration/Logging.hs [hunk 1]
```diff
@@ -211,3 +232,56 @@ createLoggingLayer ver nodeConfig' = do
...
+              -> Consensus.TopLevelConfig (CardanoBlock c)
+              -> UTCTime
+              -> IO [LogObject Text]
+nodeBasicInfo nc cfg nodeStartTime' = do
+  let CardanoLedgerConfig cfgByron cfgShelley cfgAllegra cfgMary = Consensus.configLedger cfg
+      protocolDependentItems =
+        case ncProtocol nc of
+          ByronProtocol ->
+             let genesis = byronLedgerConfig cfgByron
+             in [ ("systemStartTime",  show (getSystemStart $ Consensus.configBlock cfg))
```
Comment (denisshevchenko) [inline cardano-node/src/Cardano/Node/Configuration/Logging.hs:249]: It doesn't compile. @mrBliss could you help me please?

File: cardano-node/src/Cardano/Node/Run.hs [hunk 2]
```diff
@@ -177,19 +175,6 @@ setupTrace loggingLayer = do
...
-  -> Word64
-  -> IO ()
-traceNodeUpTime tr nodeLaunchTime = do
-  now <- getMonotonicTimeNSec
-  let upTimeInNs = now - nodeLaunchTime
-  meta <- mkLOMeta Notice Public
-  traceNamedObject tr (meta, LogValue "upTime" (Nanoseconds upTimeInNs))
-  threadDelay 1000000
-  traceNodeUpTime tr nodeLaunchTime
-
```
Comment (mrBliss) [inline cardano-node/src/Cardano/Node/Run.hs:192]: Why is this removed? I suppose because you're now tracing `nodeStartTime`? Maybe make this unrelated change in a separate commit explaining why instead of sneaking it in here.

File: cardano-node/src/Cardano/Node/Configuration/Logging.hs (hunk 1 above)
Comment (mrBliss) [reply cardano-node/src/Cardano/Node/Configuration/Logging.hs:249]: > It doesn't compile.
What's the error?

File: cardano-node/src/Cardano/Node/Configuration/Logging.hs [hunk 3]
```diff
@@ -211,3 +232,56 @@ createLoggingLayer ver nodeConfig' = do
...
+            in [ ("systemStartTime",          show (SL.sgSystemStart genesisShelley))
+               , ("slotLengthShelley",        show (mkSlotLength $ SL.sgSlotLength genesisShelley))
+               , ("slotLengthAllegra",        show (mkSlotLength $ SL.sgSlotLength genesisAllegra))
+               , ("slotLengthMary",           show (mkSlotLength $ SL.sgSlotLength genesisMary))
+               , ("epochLengthShelley",       show (SL.sgEpochLength genesisShelley))
+               , ("epochLengthAllegra",       show (SL.sgEpochLength genesisAllegra))
+               , ("epochLengthMary",          show (SL.sgEpochLength genesisMary))
+               , ("slotsPerKESPeriodShelley", show (SL.sgSlotsPerKESPeriod genesisShelley))
+               , ("slotsPerKESPeriodAllegra", show (SL.sgSlotsPerKESPeriod genesisAllegra))
+               , ("slotsPerKESPeriodMary",    show (SL.sgSlotsPerKESPeriod genesisMary))
```
Comment (mrBliss) [inline cardano-node/src/Cardano/Node/Configuration/Logging.hs:273]: There's quite a bit of duplication here. Repeating all three for all the upcoming eras is not so nice. I would write a function that takes a `ShelleyGenesis era` and a `String` or `Text` ("Shelley", "Allegra", or "Mary").

File: cardano-node/src/Cardano/Node/Configuration/Logging.hs [hunk 4]
```diff
@@ -211,3 +232,56 @@ createLoggingLayer ver nodeConfig' = do
...
+                , ("epochLengthByron", show (fromByronEpochSlots $ Gen.configEpochSlots genesis))
+                ]
+          ShelleyProtocol ->
+            let genesis = shelleyLedgerGenesis $ shelleyLedgerConfig cfgShelley
+            in [ ("systemStartTime",          show (SL.sgSystemStart genesis))
+               , ("slotLengthShelley",        show (mkSlotLength $ SL.sgSlotLength genesis))
+               , ("epochLengthShelley",       show (SL.sgEpochLength genesis))
+               , ("slotsPerKESPeriodShelley", show (SL.sgSlotsPerKESPeriod genesis))
+               ]
+          CardanoProtocol ->
```
Comment (mrBliss) [inline cardano-node/src/Cardano/Node/Configuration/Logging.hs:260]: What about Byron?

File: cardano-node/src/Cardano/Node/Configuration/Logging.hs [hunk 5]
```diff
@@ -211,3 +232,56 @@ createLoggingLayer ver nodeConfig' = do
...
+            let genesisShelley = shelleyLedgerGenesis $ shelleyLedgerConfig cfgShelley
+                genesisAllegra = shelleyLedgerGenesis $ shelleyLedgerConfig cfgAllegra
+                genesisMary    = shelleyLedgerGenesis $ shelleyLedgerConfig cfgMary
+            in [ ("systemStartTime",          show (SL.sgSystemStart genesisShelley))
```
Comment (mrBliss) [inline cardano-node/src/Cardano/Node/Configuration/Logging.hs:264]: You are using `show` here, what does the output look like? Best to make sure it doesn't look like `SlotLength { getSlotLength :: ... }` or something, because that makes parsing brittler.

**Comment by denisshevchenko** [comment]
@mrBliss The error is this one:

```
src/Cardano/Node/Configuration/Logging.hs:249:46: error:
    • Couldn't match type ‘Cardano.Ledger.Crypto.HASH c’
                     with ‘Cardano.Crypto.Hash.Blake2b.Blake2b_256’
        arising from a use of ‘getSystemStart’
```

**Comment by mrBliss** [comment]
> IMPORANT: Now it doesn't compile, I need help with Byron start time.
Best to make this a draft. That was an option when opening the PR, but you can still do it now, there's a link below the list of reviewers.

**Comment by mrBliss** [comment]
> @mrBliss The error is this one:
> 
> ```
> src/Cardano/Node/Configuration/Logging.hs:249:46: error:
>     • Couldn't match type ‘Cardano.Ledger.Crypto.HASH c’
>                      with ‘Cardano.Crypto.Hash.Blake2b.Blake2b_256’
>         arising from a use of ‘getSystemStart’
> ```

Try adding the `CardanoHardForkConstraints c` constraint. You can import it from `Ouroboros.Consensus.Cardano.CanHardFork`.

File: cardano-node/src/Cardano/Node/Run.hs (hunk 2 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/Node/Run.hs]: Fixed. It will be removed in a separate PR.

File: cardano-node/src/Cardano/Node/Configuration/Logging.hs (hunk 1 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/Node/Configuration/Logging.hs]: @mrBliss Now I see another error:

```
      Expected type: Consensus.TopLevelConfig (Consensus.CardanoBlock c)
        Actual type: Consensus.TopLevelConfig blk
```

**Comment by denisshevchenko** [comment]
@mrBliss It is ready for review (no more compile errors anymore).

**Comment by deepfire** [comment]
@denisshevchenko, can we also re-use `nodeBasicInfo` during startup, instead of triggering individual traces separately?

**Review by deepfire — CHANGES_REQUESTED** [review]
@denisshevchenko, can we also re-use nodeBasicInfo during startup, instead of triggering individual traces (for commit/release/version and other similar startup-only parameters traces) separately?

**Review by deepfire — COMMENTED** [review]
@denisshevchenko, can we also re-use `nodeBasicInfo` during startup, instead of triggering individual traces (for commit/release/version and other similar startup-only parameters traces) separately?

**Review by deepfire — APPROVED** [review]
LGTM now, thanks Denis!

File: cardano-node/src/Cardano/Node/Configuration/Logging.hs [hunk 6]
```diff
@@ -211,3 +240,51 @@ createLoggingLayer ver nodeConfig' = do
...
+       , ("slotLengthByron",  show (WCT.getSlotLength . fromByronSlotLength $ genesisSlotLength genesis))
+       , ("epochLengthByron", show (unEpochSize . fromByronEpochSlots $ Gen.configEpochSlots genesis))
+       ]
+  getGenesisValues era config =
+    let genesis = shelleyLedgerGenesis $ shelleyLedgerConfig config
+    in [ ("systemStartTime",          show (SL.sgSystemStart genesis))
+       , ("slotLength" <> era,        show (WCT.getSlotLength . WCT.mkSlotLength $ SL.sgSlotLength genesis))
+       , ("epochLength" <> era,       show (unEpochSize . SL.sgEpochLength $ genesis))
+       , ("slotsPerKESPeriod" <> era, show (SL.sgSlotsPerKESPeriod genesis))
+       ]
```
Comment (intricate) [inline cardano-node/src/Cardano/Node/Configuration/Logging.hs:290]: Is it appropriate to just use `show` on these values? What does the output look like?

I'd much prefer `render*` functions that ensure these values are appropriately formatted and rendered to `Text`.

**Review by intricate — APPROVED** [review]
If the text representation of the values in `nodeBasicInfo` looks okay, then we can merge this.

File: cardano-node/src/Cardano/Node/Configuration/Logging.hs (hunk 6 above)
Comment (denisshevchenko) [reply cardano-node/src/Cardano/Node/Configuration/Logging.hs:290]: > If the text representation of the values in nodeBasicInfo looks okay, then we can merge this.

@intricate Yes, the representation of these values is okay: it's the simplest possible. :-) For example, `slotLength` will be rendered not as:

```
SlotLength { getSlotLength = 1 }
```

but as:

```
1
```
